app/models/user.rb

```diff
@@ -0,0 +1,22 @@
+class User < ApplicationRecord
+  before_save :downcase_email
+
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
+
+  validates :email, presence: true,
+    length: {maximum: Settings.user.LENGTH_EMAIL_MAX},
+    format: {with: VALID_EMAIL_REGEX},
+    uniqueness: {case_sensitive: false}
+  validates :name, presence: true,
+    length: {maximum: Settings.user.LENGTH_NAME_MAX}
+  validates :password, presence: true,
+    length: {minimum: Settings.user.LENGTH_PASSWORD_MIN}
+
+  has_secure_password
```

Comment:
tìm hiểu và nói cho mình lệnh trên có tác dụng gì nhé?

Reply:
- Cung cấp khả năng lưu thuộc tính băm an toàn password_digest (tương tự với khái niệm bản mã trong mã hóa dữ liệu) tới cơ sở dữ liệu bằng cách sử dụng hàm băm bcrypt.
- Kiểm tra đầu vào và kiểm tra trùng khớp cho cặp thuộc tính ảo password và password_confirmation
- Thông báo lại người dùng khi mật khẩu là đúng hoặc sai.

Reply:
kỹ hơn chút về phần validate nhé

Reply:
- Xác định trường hợp nhập mật khẩu trống và thiết lập độ dài tối thiểu của mật khẩu.
- Mật khẩu người dùng nhập vào sẽ qua hàm băm bcrypt để cho ra một giá trị băm. Lấy giá trị này so sánh với password_digest, nếu giống nhau thì pass validate password.

Reply:
nó sẽ validate trường hợp không nhập mật khẩu khi nào?

Reply:
Khi user mới được tạo, khi đăng nhập và khi đổi mật khẩu tới mật khẩu mới.

Reply:
khi đăng nhập thì model nó validate như thế nào?

Reply:
À, mình nhầm trường chỉ có 2 trường hợp tạo mới và thay đổi mật khẩu mới validate trường hợp không nhập mật khẩu.

Reply:
thật ra thì trường hợp thay đổi mật khẩu cái validate đó không phải do has_secure_password validate đâu.